### PR TITLE
S3 lambda sqs

### PR DESCRIPTION
**Pull request by cyga91**

### CODE REVIEW

File: src/main/java/helloworld/HelloWord.java [hunk 1]
```diff
@@ -8,19 +8,25 @@
...
+import static helloworld.constant.Constants.LAMBDA_FUNCTION_OUTPUT_HANDLER;
 import static helloworld.constant.Constants.LAMBDA_FUNCTION_OUTPUT_NAME;
+import static helloworld.constant.Constants.LAMBDA_FUNCTION_ROLE;
 import static helloworld.constant.Constants.QUEUE_OUTPUT_NAME;
 
 public class HelloWord {
     public static void main(String[] args) {
         S3Bucket bucket = new S3Bucket(AmazonS3ClientBuilder.standard().withRegion(Regions.EU_WEST_1).build(), BUCKET_NAME_OUTPUT);
         Sqs output_queue = new Sqs(AmazonSQSClientBuilder.defaultClient(), QUEUE_OUTPUT_NAME);
+        Lambda lambdaOutput = Lambda.builder().lambdaName(LAMBDA_FUNCTION_OUTPUT_NAME).role(LAMBDA_FUNCTION_ROLE).s3BucketName(BUCKET_NAME_OUTPUT).s3BucketKey(BUCKET_OUTPUT_KEY).handler(LAMBDA_FUNCTION_OUTPUT_HANDLER).build();
```
Comment (amsoares) [inline src/main/java/helloworld/HelloWord.java]: With method chaining, when there are many of them (more than 120 characters, the limit that your IntelliJ shows a line in the vertical), it's a good practice to break them down. It improves a lot the readability ;) 
```suggestion
        Lambda lambdaOutput = Lambda.builder()
                .lambdaName(LAMBDA_FUNCTION_OUTPUT_NAME)
                .role(LAMBDA_FUNCTION_ROLE)
                .s3BucketName(BUCKET_NAME_OUTPUT)
                .s3BucketKey(BUCKET_OUTPUT_KEY)
                .handler(LAMBDA_FUNCTION_OUTPUT_HANDLER)
                .build();
```

File: src/main/java/helloworld/s3_lambda_sqs/ReadWriteSQSLambdaAPI.java [hunk 2]
```diff
@@ -4,19 +4,43 @@
...
+        client.send(request, HttpResponse.BodyHandlers.ofString());
+        logger.log(MESSAGE_SENT_TO_API + request.toString());
+
         return "200 Ok";
     }
+
+    private HttpRequest createSendHttpRequest(String result) {
+        return HttpRequest.newBuilder()
+                .uri(URI.create(POSTS_API_URL))
+                .header("accept", "application/json")
```
Comment (amsoares) [inline src/main/java/helloworld/s3_lambda_sqs/ReadWriteSQSLambdaAPI.java]: You can have these headers and values as constants as well

File: src/main/java/helloworld/s3_lambda_sqs/ReadWriteS3LambdaSQS.java [hunk 3]
```diff
@@ -4,15 +4,32 @@
...
+        String result = null;
 
         for (SQSEvent.SQSMessage msg : sqsEvent.getRecords()) {
-            logger.log(msg.getBody());
+            result = msg.getBody();
+            logger.log(SQS_RECEIVED_MESSAGE + msg.getBody());
         }
-        return null;
+
+        Objects.requireNonNull(result);
```
Comment (amsoares) [inline src/main/java/helloworld/s3_lambda_sqs/ReadWriteS3LambdaSQS.java]: Although this works fine here, it should be used in the context of `Fail-fast` checks, and not as a flow control. For instance, when creating an object through constructor, or when passing some parameters to your method and you want not to process further. 

Here we also assume that AWS Event won't ever come with body==null, so I'd suggest that if you want to double check, or you throw the exception from the for loop above, or checking it now for nullity.

File: src/main/java/helloworld/s3_lambda_sqs/ReadWriteS3LambdaSQS.java [hunk 4]
```diff
@@ -4,15 +4,32 @@
...
     @Override
     public String handleRequest(SQSEvent sqsEvent, Context context) {
         LambdaLogger logger = context.getLogger();
+        Sqs queue = new Sqs(AmazonSQSClientBuilder.defaultClient(), QUEUE_INPUT_NAME);
+
+        String result = null;
 
         for (SQSEvent.SQSMessage msg : sqsEvent.getRecords()) {
-            logger.log(msg.getBody());
+            result = msg.getBody();
```
Comment (amsoares) [inline src/main/java/helloworld/s3_lambda_sqs/ReadWriteS3LambdaSQS.java]: So far AWS sends Records array always with a single item. Just for FYI
https://stackoverflow.com/questions/40765699/how-many-records-can-be-in-s3-put-event-lambda-trigger/40767563#40767563

File: src/main/java/helloworld/constant/Constants.java [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package helloworld.constant;
+
+public final class Constants {
```
Comment (amsoares) [inline src/main/java/helloworld/constant/Constants.java:5]: A class, in order to be a truly immutable class, should also define a private constructor, preventing its instantiation. Was it the idea?

Ref: http://www.javapractices.com/topic/TopicAction.do?Id=2

File: src/main/java/helloworld/s3_lambda_sqs/SQSMessageSender.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+import java.util.logging.Handler;
+
+import static helloworld.constant.Constants.MESSAGE;
+import static helloworld.constant.Constants.MESSAGE_SENT;
+import static helloworld.constant.Constants.QUEUE_NAME;
+import static helloworld.constant.Constants.WITH_DELAY_SECONDS;
+
+public class SQSMessageSender {
+    Gson gson = new GsonBuilder().setPrettyPrinting().create();
```
Comment (amsoares) [inline src/main/java/helloworld/s3_lambda_sqs/SQSMessageSender.java]: if the intention is to have a constant variable, please declare it as static final ;)

File: src/main/java/helloworld/s3_lambda_sqs/SQSMessageSender.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    private static final Logger logger = LoggerFactory.getLogger(Handler.class);
+
+    public static void main(String[] args) {
+        final AmazonSQS sqs = AmazonSQSClientBuilder.defaultClient();
+
+        // Create queue if doesn't exists
+        try {
+            CreateQueueResult create_result = sqs.createQueue(QUEUE_NAME);
```
Comment (amsoares) [inline src/main/java/helloworld/s3_lambda_sqs/SQSMessageSender.java]: I'd suggest keeping the camelCase convention :)
```suggestion
            CreateQueueResult createResult = sqs.createQueue(QUEUE_NAME);
```

File: src/main/java/helloworld/s3_lambda_sqs/SQSMessageSender.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        } catch (AmazonSQSException e) {
+            if (!e.getErrorCode().equals("QueueAlreadyExists")) {
```
Comment (amsoares) [inline src/main/java/helloworld/s3_lambda_sqs/SQSMessageSender.java]: Is there any special reason for not logging every exception?

File: src/main/java/helloworld/s3_lambda_sqs/SQSMessageSender.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+                logger.error(e.getMessage());
+            }
+        }
+
+        String queueUrl = sqs.getQueueUrl(QUEUE_NAME).getQueueUrl();
+
+        // Send multiple messages to the queue
+        SendMessageBatchRequest send_batch_request = new SendMessageBatchRequest()
```
Comment (amsoares) [inline src/main/java/helloworld/s3_lambda_sqs/SQSMessageSender.java]: camelCase welcomed here :)

File: pom.xml [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        <!--loggs-->
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-api</artifactId>
+            <version>2.13.0</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-core</artifactId>
+            <version>2.13.2</version>
```
Comment (amsoares) [inline pom.xml]: Maybe you could parameterize this versions? Also would be nice to stick to same version (if it's possible, of course)

File: src/main/java/helloworld/s3_lambda_sqs/S3Bucket.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        final AmazonS3 s3 = AmazonS3ClientBuilder.standard().withRegion(Regions.EU_WEST_1).build();
+
+        Bucket bucket = createBucket();
+        Objects.requireNonNull(bucket);
+
+        try {
+            s3.putObject(BUCKET_NAME, KEY_NAME, new File(FILE_PATH));
+        } catch (AmazonServiceException e) {
+            logger.error(e.getErrorMessage());
+            System.exit(1);
```
Comment (amsoares) [inline src/main/java/helloworld/s3_lambda_sqs/S3Bucket.java]: `System.exist(1)` should be used in contexts like described here: https://stackoverflow.com/a/3716015
In the case of Lambdas, we can just throw the exception up, and the AWS Lamda engine takes care of the rests.

Also another important point to consider is that, if System.exit would kill our program as it's intended with System.exit, we would loose the context sharing of our Lambda, leading to more often Cold Starts. Some interesting material about it here: https://aws.amazon.com/blogs/compute/new-for-aws-lambda-predictable-start-up-times-with-provisioned-concurrency/

File: src/main/java/helloworld/s3_lambda_sqs/S3Bucket.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        }
+        logger.info(SUCCESS_WRITE_TO_S3);
+    }
+
+    public static Bucket createBucket() {
+        final AmazonS3 s3 = AmazonS3ClientBuilder.standard().withRegion(Regions.EU_WEST_1).build();
```
Comment (amsoares) [inline src/main/java/helloworld/s3_lambda_sqs/S3Bucket.java]: I see that this guy here has been created in other methods. Maybe is a good idea to set it as instance variable? Such way you don't need to instantiate it in every method.

File: src/main/java/helloworld/s3_lambda_sqs/S3Bucket.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+        logger.info(CURRENT_BUCKET + BUCKET_NAME);
+        return b;
+    }
+
+    public static Bucket getBucket(String bucketName) {
+        final AmazonS3 s3 = AmazonS3ClientBuilder.standard().withRegion(Regions.EU_WEST_1).build();
+
+        Bucket bucket = null;
+        List<Bucket> buckets = s3.listBuckets();
+        for (Bucket b : buckets) {
```
Comment (amsoares) [inline src/main/java/helloworld/s3_lambda_sqs/S3Bucket.java]: I'd suggest using the Java Stream API. The flow control is handled better by the JVM, and we reduce verbosity.
Also, any nullity or missing bucket will be under the Stream control. Look and let me know what you think.

```suggestion
        return s3.listBuckets().stream()
                .filter(bucket -> bucket.getName().equals(bucketName))
                .findFirst()
                .orElseThrow(new Exception("Cant find the bucket"));
```

Comment (amsoares) [reply src/main/java/helloworld/s3_lambda_sqs/S3Bucket.java]: Alternatively you could use the forEach
```suggestion
        s3.listBuckets().forEach(eachBuckett -> /* your logic here */);
```

File: src/main/java/helloworld/s3_lambda_sqs/Lambda.java [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+            awsLambda.createFunction(functionRequest);
+        } catch (ServiceException e) {
+            logger.error(e.getErrorMessage());
+            System.exit(1);
+        }
+        checkLambdaFunction(lambdaFunctionName);
+    }
+
+    private static CreateFunctionRequest getFunctionRequest(String lambdaFunctionName) {
+        CreateFunctionRequest functionRequest = new CreateFunctionRequest();
```
Comment (amsoares) [inline src/main/java/helloworld/s3_lambda_sqs/Lambda.java]: It's always good to use the Builder ;)

File: src/main/java/helloworld/HelloWord.java (hunk 1 above)
Comment (cyga91) [reply src/main/java/helloworld/HelloWord.java]: what u mean?

**Pull request merged**